### DISCUSSION TITLE
A WARNING with ADKernel

### DISCUSSION THREAD

**Discussion by wangzhaohao** [idaholab/moose#28527] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi, there is a custom *Diffusion* Kernel. Seems all is right, because it's quite simple.
```C++
#pragma once

#include "GenericKernel.h"

template <bool is_ad>
class OxygenDiffusionTempl : public GenericKernel<is_ad>
{
public:
  static InputParameters validParams();
  OxygenDiffusionTempl(const InputParameters & parameters);

protected:
  virtual GenericReal<is_ad> computeQpResidual() override;
  virtual Real computeQpJacobian() override;

  const VariableValue & _temperature; //[K]
  usingGenericKernelMembers;
  using GenericKernel<is_ad>::_grad_test;
  using GenericKernel<is_ad>::_grad_u;
  using GenericKernel<is_ad>::_grad_phi;
private:
  Real _oxygen_diffusion_coeff; //[m^2/s]
};

typedef OxygenDiffusionTempl<false> OxygenDiffusion;
typedef OxygenDiffusionTempl<true> ADOxygenDiffusion;
```

```C++
#include "OxygenDiffusion.h"

registerMooseObject("MiracleApp", OxygenDiffusion);
registerMooseObject("MiracleApp", ADOxygenDiffusion);

template <bool is_ad>
InputParameters OxygenDiffusionTempl<is_ad>::validParams()
{
  InputParameters params = GenericKernel<is_ad>::validParams();
  params.addCoupledVar("temperature", "the temperature of the MOX or UO2 fuel in K");
  params.addClassDescription("Compute the oxygen diffusion term in MOX fuel.");
  return params;
}

template <bool is_ad>
OxygenDiffusionTempl<is_ad>::OxygenDiffusionTempl(const InputParameters & parameters)
: GenericKernel<is_ad>(parameters),
_temperature(this->coupledValue("temperature"))
{
}

template <bool is_ad>
GenericReal<is_ad> OxygenDiffusionTempl<is_ad>::computeQpResidual()
{
  _oxygen_diffusion_coeff = 3.5e-9 * std::exp(-7500. / _temperature[_qp]);// modifyed according FEAST_OXIDAL paper  
  return _oxygen_diffusion_coeff * _grad_test[_i][_qp] * _grad_u[_qp];
}

template <bool is_ad>
Real OxygenDiffusionTempl<is_ad>::computeQpJacobian()
{
  if constexpr (!is_ad)
  return _oxygen_diffusion_coeff * _grad_test[_i][_qp] * _grad_phi[_j][_qp];
  else
  mooseError("In ADOxygenDiffusion, computeQpJacobian should  not be called!");
}

template class OxygenDiffusionTempl<false>;
template class OxygenDiffusionTempl<true>;
```

The Miracle is myself AppName, when I test the Kernel, with this input file.
```Assembly
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 10
  ny = 10
[]

[Variables]
  [u]
  []
[]
[AuxVariables/temperature]
  initial_condition = 1000
[]

[AuxKernels]
  [temperature]
    type = ConstantAux
    variable = temperature
    value = 1000
  []
[]

[Kernels]
  [diff]
    type = ADOxygenDiffusion
    variable = u
    temperature = temperature
  []
[]

[BCs]
  [left]
    type = ADDirichletBC
    variable = u
    boundary = left
    value = 0
  []
  [right]
    type = ADDirichletBC
    variable = u
    boundary = right
    value = 1
  []
[]

[Executioner]
  type = Transient
  dt = 1
  end_time = 10
  solve_type = 'PJFNK'
#  petsc_options_iname = '-pc_type -pc_hypre_type'
#  petsc_options_value = 'hypre boomeramg'
  nl_abs_tol = 1e-10
[]

[Outputs]
  exodus = true
[]
```
there is a WARNING 
```Shell
WARNING! There are options you set that were not used!
WARNING! could be spelling mistake, etc!
There is one unused database option. It is:
Option left: name:-ksp_converged_reason value: ::failed source: code
```
Althought There output exodus looks right, I want to remove this WARNING. So could you give me some help, Thanks!!!

Zhaohao Wang

**Comment by wangzhaohao** [comment]
There is my MOOSE Version.
```Shell
MOOSE Version:           git commit dfb9763289 on 2024-07-15
LibMesh Version:         
PETSc Version:           3.20.3
SLEPc Version:           3.20.1
Current Time:            Tue Sep  3 10:46:54 2024
Executable Timestamp:    Tue Sep  3 10:46:42 2024
```

and my workstation is 

 ██████████████████  ████████     wang@wang-x11dain
 ██████████████████  ████████     OS: Manjaro 24.0.7 Wynsdey
 ██████████████████  ████████     Kernel: x86_64 Linux 6.5.13-7-MANJARO
 ██████████████████  ████████     Uptime: 8d 2h 3m
 ████████                             ████████     Packages: 1679
 ████████  ████████     ████████     Shell: zsh 5.9
 ████████  ████████     ████████     Resolution: 3000x2560
 ████████  ████████     ████████     DE: KDE
 ████████  ████████     ████████     WM: KWin
 ████████  ████████     ████████     GTK Theme: Breeze [GTK2/3]
 ████████  ████████     ████████     Icon Theme: Tela
 ████████  ████████     ████████     Disk: 161G / 505G (34%)
 ████████  ████████     ████████     CPU: Intel Xeon Gold 6148 @ 80x 3.7GHz [37.0°C]
 ████████  ████████     ████████     GPU: NVIDIA GeForce GTX 1050
                                                                              RAM: 18681MiB / 96296MiB

**Comment by GiudGiud** [comment, marked as answer]
Thank you for reporting. 
@pbehne has been removing superfluous petsc options warnings / making it as accurate as possible. He will be able to help you though there might be a delay as the end of the financial year is a busier period

In the meantime, this is just a warning and you can ignore it.

**Comment by pbehne** [comment]
Are there any nonlinear iterations in your solve? If not, you are likely encountering [this](https://github.com/idaholab/moose/issues/27859). This issue has been fixed in PETSc, we are just waiting for the next petsc submodule update.

If you MUST turn the warning off, you can supply `-options-left 0` on the command line. I wouldn't recommend doing this, as it will turn off ALL warnings about unused options. But, if you are sure, go for it.
